File: Helmut/9/solve.jl
function follow(t, h)
    d = h - t
    tt = t
    if maximum(abs.(d))>1 
        if abs(d[1]) > 0
            tt[1] += div(d[1], abs(d[1]))
        end
        if abs(d[2]) > 0
            tt[2] += div(h[2] - t[2], abs(h[2] - t[2]))
        end
    end
    tt
end

function run(fn, lenSnake)
    m = readlines(open(fn))
    h = collect([0,0] for _ in 1:lenSnake)
    seen = Set()
    push!(seen, tuple(h[lenSnake]...))
    dd = Dict("R" => [1, 0], "L" => [-1, 0], "U" => [0, 1], "D" => [0, -1])
    for (c, l) in split.(m)
        for iter in 1:parse(Int, l)
            h[1] += dd[c]
            for s in 2:lenSnake
                h[s] = follow(h[s],h[s-1])
            end
            push!(seen, tuple(h[end]...))
         end
    end
    println(length(seen))
end

run("small.txt",2)
run("input",2)
run("medium.txt",10)
run("input",10)

File: Helmut/9/solve.hs
import Data.Set (fromList) 
import Control.Monad (zipWithM) 

type Pos = (Int, Int)   -- ein Feld (x,y)
type Move = (Int, Int)  -- eine Richtung (dx, dy)
type Schlange = [Pos]   -- die Segmente als Felder
type Moves = [Move]     -- eine Folge von Richtungen

-- Segment (c,d) folgt Segment (a,b)
follow :: Pos -> Pos -> Pos
follow (a,b) (c,d) = if max (abs (a-c)) (abs (b-d)) <= 1 
                         then (c,d)
                         else (f a c, f b d)
            where f x y = y + (signum (x-y))

move1step :: Schlange -> Move -> Schlange
move1step ((x,y):snake) (dx,dy) = scanl follow (x+dx, y+dy) snake

parse :: String -> Moves
parse (move:n) =  let dir = case move of { 
                      'L' -> (-1,0) ; 'R' -> (1,0) ; 'U' -> (0,1) ; 'D' -> (0,-1) }
                   in replicate (read n) dir

-- durch das fromList hat man alle Felder nur einmal
moveAndCount :: Schlange -> Moves -> Int
moveAndCount schlange = length . fromList . map last . scanl move1step schlange 

kurzeSchlange = replicate 2 (0,0)  :: Schlange
langeSchlange = replicate 10 (0,0) :: Schlange

test1 = concatMap parse ["R 4", "U 4"]       :: Moves
test2 = move1step [(0,0),(0,0)] (1,0)        :: Schlange
test3 = foldl move1step kurzeSchlange test1
test4 = foldl move1step langeSchlange test1

main = zipWithM run [kurzeSchlange, langeSchlange, langeSchlange] 
                    ["small.txt", "medium.txt", "input"]
       where run schlange fn = fmap (moveAndCount schlange . concatMap parse . lines) 
                                  $ readFile fn
   
